Add more verbose error messages for incorrect or missing signoffs

hiimbex opened the pull request:
The first 2 commits of this PR update the `.gitignore` to not include the `.env` file and update to probot 0.8.0 and use the latest standards (ie `(event, context)` => `context`, integration => app).

This PR adds more verbose error messages if someone does not have a DCO sign off. it modifies `lib/dco.js` to return true if there is a sign off or an error message from the `getError()` method. It then checks if a string was returned and if so it adds that to the `defaults.failure.description` (checking to make sure it isn't already added in the case of multiple commits with incorrect sign offs). 

Closes #18. @bkeepers I really didn't even think that the bot would be running on its own repo and I dug myself a bit of a hole lol

bkeepers commented on lib/dco.js:
```diff
@@ -1,8 +1,28 @@
+    err = 'Your author name does not match. ';
+  } else if (!commit.message.includes(commit.author.email)) {
+    err = 'Your email does not match. ';
+  }
```
It feels like there's some duplication between the `getError` method and the exported function at the top. I wonder if it could be reduced by replacing the `signOff` method with a regular expression that includes matching groups for the author name and email. Something like this (I haven't tested this, so will likely need tweaked): 

```js
const regex = /^Signed-off-by: (.*) <(.*)>$/gm
const [match, name, email] = commit.message.match(regex);

if (!match) {
  return 'The sign-off is missing.';
} else if (commit.author.name !== name || commit.author.email !== email) {
  return `The sign-off does not match the commit author: expected "${commit.author.name} <${commit.author.email}>", but got "${name} <${email}>"';
}
```

hiimbex replied:
So the only problem with this approach is the literal message length. So the message (as far as I know) only appears in 2 places on a pull request, at the bottom under checks, where it gets cut of past that one line, and clicking details redirects [here](https://github.com/probot/dco/blob/master/index.js#L11). The other place it shows up is if you hover over the X or ✅ it displays the message; _however_ if you have multiple checks (like here there's 2, the dco and travis CI) then it displays 1/2 checks OK.

So I'm 100% for this solution, in fact mine has a minor bug? ie my expected name and email were `hiimbex` and `hiimbex@github.com` and I tested with `othername` and `hiimbex@github.com` and the matching I was doing would count that as correct when it shouldn't have, but the ability to not view the error seems pretty critical.

hiimbex commented:
@bkeepers 💯 for linking how to fix things as well. I'd love for this bot to not comment and be as simple as setting a status, but there seems to be a lot of information we need to convey 🤔

hiimbex commented:
@bkeepers I implemented the changes we discussed using the regex. One interesting result is since someone could have multiple incorrectly authored/emailed sign-offs (I had many for testing purposes), I added a check to only include the first incorrect one [here](https://github.com/probot/dco/pull/19/files#diff-168726dbe96b3ce427e7fedce31bb0bcR35). This prevents us from going over the API defined limit, which interestingly enough is only [1024 bytes](https://developer.github.com/v3/repos/statuses/#create-a-status).

I updated `target_url` to redirect to https://github.com/probot/dco/#how-it-works. I also think this PR can be merged in on its own, without updating that section. And, considering I have been unsuccessful in rebasing and actually singing off on all these commits, I don't think I'm particularly qualified to write that section 😅

hiimbex commented:
Memorializing a conversation with @bkeepers, we're gonna pull the logic out from `lib/dco.js` for improved readability and to avoid checking if the method returns a string

hiimbex commented:
code updated and dco finally fixed but not sure what to do about the tests 🙇‍♀️ 

cc/ @bkeepers

hiimbex commented:
Moved back to an extracted function that would return the DCO object that is passed into the GitHub API. For testing I compared `JSON.stringify`'d versions of the objects to verify they contained the correct content (not sure if there's a better way to verify that two objects have the same sets of property names and values).

(You'd think @bkeepers would get the hint by the 5th comment 😜🙏🏼)

jsquyres commented:
Sweet -- thanks for doing this!

Does this mean that if we're using the github-hosted DCO app integration, we get this new feature automatically?

hiimbex commented:
@jsquyres yes! It is completely automatic. You should see it in action the next time a sign off fails ☺️
